lib/concurrent/promise.rb

```diff
@@ -438,6 +438,46 @@ def self.any?(*promises)
       aggregate(:any?, *promises)
     end
 
+    # Aggregates a collection of promises, and completes as soon as the first
+    # aggregated promise completes, with the same state (fulfilled or rejected)
+    # and reason/value.
+    #
+    #   @param [Array] promises Zero or more promises to aggregate
+    #   @return [Promise] an unscheduled (not executed) promise that aggregates
+    #     the promises given as arguments
+    def self.race(*promises)
+      return Promise.fulfilled if promises.empty?
+
+      Promise.new do
+        latch = Concurrent::CountDownLatch.new(1)
+        finished = Concurrent::AtomicBoolean.new(false)
```

Comment:
I don't think it currently has any race conditions, it's really the AtomicBoolean alone that guards 'first to finish' access, from which everything else follows. 

But I catch your point, I agree it would be better to use one single construct to control everything, I just didn't see a way to do so.